### ISSUE TITLE
What is the best practice to set `shuffle=False` in DataLoader?

### ISSUE THREAD

**Issue by WuZhuoran** [ultralytics/ultralytics#19468]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I tried to enable `rect=True` features in yolov11 Training mode. And it just shows: 

```
WARNING `rect=True` is incompatible with DataLoader shuffle, setting shuffle=False
```

Then I found that `shuffle` is not an argument for training configuration. https://docs.ultralytics.com/modes/train/#train-settings

Then I checked the code, the shuffle is automaticly set as `True` when in training mode. https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/detect/train.py#L50

Then I checked it seems the only way is to define own DataLoader? 

What is the best practice to do so? Is there any code available for reference?

Thanks

**Comment by UltralyticsAssistant** [comment]
👋 Hello @WuZhuoran, thank you for your interest in Ultralytics 🚀!

We recommend taking a look at the [Ultralytics Documentation](https://docs.ultralytics.com/), where you may find valuable insights on training settings such as [`rect`](https://docs.ultralytics.com/modes/train/#train-settings), `DataLoader`, and other configurable options.

To address your question, if this is a ❓ **Custom Training Question**, please provide as much additional context as possible, including specific configurations, dataset examples, or any available logs. This helps us better understand your situation and guide you accordingly.

If this is a 🐛 **Bug Report**, please ensure that you've provided a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) to make the debugging process clearer and faster for us.

Additionally, make sure you’re using the latest version of the `ultralytics` package to rule out any issues that may already be resolved. You can upgrade with the following command:

```bash
pip install -U ultralytics
```

## Community and Resources

For more in-depth discussions or real-time support:
- Join our [Discord community](https://discord.com/invite/ultralytics) 🎧
- Explore [Ultralytics Discourse](https://community.ultralytics.com/) for Q&A
- Share insights and learn via our [Subreddit](https://reddit.com/r/Ultralytics)

## Verified Environments

Testing your code in one of the verified environments can simplify setup and help isolate potential issues:
- **Notebooks** with GPU 🖥️: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud setups** 🌥️: [GCP](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/) VMs
- **Docker** 🐋: See the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status Check

To ensure all systems are functioning correctly, refer to our CI badge below:
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

An Ultralytics engineer will review this soon and assist further—thank you for your patience 😊!

**Comment by Y-T-G** [comment]
You can't change it without modifying code.

**Comment by WuZhuoran** [comment]
> You can't change it without modifying code.

Thanks for reply! So the only way is to modify the source code? Can I load a customized DataLoader to override this config? 

And how to enable "rect" config without modifying code? There should be a way otherwise this "rect" config will be pretty hard to use.

Thank you

**Comment by Y-T-G** [comment]
You can pass `rect` to model.train()

To override the dataloader, you will need to override the trainer.

**Comment by WuZhuoran** [comment]
> You can pass `rect` to model.train()
> 
> To override the dataloader, you will need to override the trainer.

If we used `rect` in model.train(), it will show: `WARNING ⚠️ 'rect=True' is incompatible with DataLoader shuffle, setting shuffle=False`

Is there a way to use `rect` in model.train() without modifying the original code?

**Comment by Y-T-G** [comment]
`rect` doesn't support shuffle even if you modify the code

**Comment by WuZhuoran** [comment]
> `rect` doesn't support shuffle because it sorts image based on aspect ratio so that it can create batches with minimum padding 

Thanks. I understand why it does not allow shuffle. I just want to know how to run yolo train with "rect" True and without modifying the source code?

Such as "yolo train ... rect=True" something. 

If this is a config in training but it still need modifying source code, how this config been used by users?

Thanks,

**Comment by Y-T-G** [comment]
You use `rect=True` in model.train() to use it. It will disable shuffle. But rect works.

**Comment by WuZhuoran** [comment]
> You use `rect=True` in model.train() to use it. It will disable shuffle. But rect works.

I did but during the training it still showed:

```
WARNING ⚠️ updating to 'imgsz=640'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'
WARNING ⚠️ 'rect=True' is incompatible with DataLoader shuffle, setting shuffle=False
```

And here is my config:

```
epochs: 600  # number of epochs to train for
patience: 100  # epochs to wait for no observable improvement for early stopping of training
batch: 24  # number of images per batch (-1 for AutoBatch)
imgsz: [640, 360] # [1024,576]  # size of input images as integer or w,h
save: True  # save train checkpoints and predict results
save_period: -1 # Save checkpoint every x epochs (disabled if < 1)
cache: False  # True/ram, disk or False. Use cache for data loading
device: 1 # device to run on, i.e. cuda device=0 or device=0,1,2,3 or device=cpu
workers: 8  # number of worker threads for data loading (per RANK if DDP)
project: # project name
name: 'v11x_imgsz_640_360_aug' # experiment name, results saved to 'project/name' directory
exist_ok: False  # whether to overwrite existing experiment
pretrained: True  # whether to use a pretrained model
optimizer: Adam  # optimizer to use, choices=['SGD', 'Adam', 'AdamW', 'RMSProp']
verbose: True  # whether to print verbose output
seed: 19960214  # random seed for reproducibility
deterministic: True  # whether to enable deterministic mode
single_cls: False  # train multi-class data as single-class
rect: True  # support rectangular training if mode='train', support rectangular evaluation if mode='val'
cos_lr: True  # use cosine learning rate scheduler
close_mosaic: 0  # disable mosaic augmentation for final 10 epochs
resume: False  # resume training from last checkpoint
# Segmentation
overlap_mask: True  # masks should overlap during training (segment train only)
mask_ratio: 4  # mask downsample ratio (segment train only)
# Classification
dropout: 0.2  # use dropout regularization (classify train only)
augment: True # Use Data Augmentation
```

**Comment by Y-T-G** [comment]
I don't understand. It sets `shuffle=False` because it's not compatible with rect. But rect is still being used.

**Comment by WuZhuoran** [comment]
> I don't understand. It sets `shuffle=False` because it's not compatible with rect. But rect is still being used.

Thanks. That makes sense. I thought this warning means `rect` is not applied.

**Comment by glenn-jocher** [comment]
**Answer:** The warning simply informs you that `shuffle` is automatically disabled when using `rect=True`, but rectangular training remains fully active. This is normal behavior - no action needed. For implementation details, see the [`get_dataloader` function](https://docs.ultralytics.com/reference/models/yolo/detect/train/#ultralytics.models.yolo.detect.train.DetectionTrainer.get_dataloader) in our source code.
